pkg/operator/controller/externaldns/credentials-request_test.go

```diff
@@ -0,0 +1,408 @@
+		},
+		{
+			name:                      "Ensure Credential request when invalid credential requests are there in GCP",
+			existingObjects:           []runtime.Object{getUnDesiredCredentialRequest(testGCPExtDNSInstanceWhenSourceIsOCPRoute())},
+			inputExtDNS:               testGCPExtDNSInstanceWhenSourceIsOCPRoute(),
+			expectedCredentialRequest: getDesiredCredentialRequest(testGCPExtDNSInstanceWhenSourceIsOCPRoute()),
+		},
+	}
+	for _, tc := range testCases {
+		cl := fake.NewClientBuilder().WithScheme(test.Scheme).WithRuntimeObjects(tc.existingObjects...).Build()
+
+		r := &reconciler{
+			client: cl,
+			scheme: test.Scheme,
+			config: testConfig(),
+			log:    zap.New(zap.UseDevMode(true)),
+		}
+
+		t.Run(tc.name, func(t *testing.T) {
+
+			exists, got, err := r.ensureExternalCredentialsRequest(context.TODO(), tc.inputExtDNS)
+
+			if err != nil {
+				t.Log("Error while ensuring Credentials request")
+			}
+
+			if !exists {
+				t.Errorf("Credential request does not exist")
+			}
+
+			if got.Name != tc.expectedCredentialRequest.Name {
+				t.Errorf("Got name %v  but expected name is %v", got.Name, tc.expectedCredentialRequest.Name)
+			}
+
+			if got.Namespace != tc.expectedCredentialRequest.Namespace {
+				t.Errorf("Got namespace %v  but expected namespace is %v", got.Namespace, tc.expectedCredentialRequest.Namespace)
+			}
+
+			if tc.inputExtDNS.Spec.Provider.Type == operatorv1alpha1.ProviderTypeAWS {
+				gotDecodedAWSSpec, expectedAWSSpec := decodeAWSProviderSpec(*got, *tc.expectedCredentialRequest)
+				if !reflect.DeepEqual(gotDecodedAWSSpec, expectedAWSSpec) {
+					t.Errorf("Got CredentialRequest Spec %v\n  but expected CredentialRequest Spec is %v", gotDecodedAWSSpec, expectedAWSSpec)
+				}
+
+			}
+
+			if tc.inputExtDNS.Spec.Provider.Type == operatorv1alpha1.ProviderTypeGCP {
+				gotDecodedGCPSpec, expectedGCPSpec := decodeGCPProviderSpec(*got, *tc.expectedCredentialRequest)
+				if !reflect.DeepEqual(gotDecodedGCPSpec, expectedGCPSpec) {
+					t.Errorf("Got CredentialRequest Spec %v\n  but expected CredentialRequest Spec is %v", gotDecodedGCPSpec, expectedGCPSpec)
+				}
+
+			}
+
+			if tc.inputExtDNS.Spec.Provider.Type == operatorv1alpha1.ProviderTypeAzure {
+				gotDecodedAzureSpec, expectedAzureSpec := decodeAzureProviderSpec(*got, *tc.expectedCredentialRequest)
+				if !reflect.DeepEqual(gotDecodedAzureSpec, expectedAzureSpec) {
+					t.Errorf("Got CredentialRequest Spec %v\n  but expected CredentialRequest Spec is %v", gotDecodedAzureSpec, expectedAzureSpec)
+				}
+
+			}
+
+		})
+
+	}
+
+}
+
+func decodeGCPProviderSpec(gotCrededentialrequest, expectedCredentialRequest cco.CredentialsRequest) (gotDecodedGCPSpec, expectedDecodedGCPSpec cco.GCPProviderSpec) {
+
+	codec, _ := cco.NewCodec()
+	gotDecodedGCPSpec = cco.GCPProviderSpec{}
+	err := codec.DecodeProviderSpec(gotCrededentialrequest.Spec.ProviderSpec, &gotDecodedGCPSpec)
+	if err != nil {
+		return gotDecodedGCPSpec, cco.GCPProviderSpec{}
```

Comment:
I'm still not quite convinced that these 2 cannot be the same. Return an error if the decode fails and check it in the test, this would avoid any possible comparison mistake.